cores/esp8266/core_esp8266_postmortem.cpp

```diff
@@ -322,5 +403,199 @@ void __stack_chk_fail(void) {
+    ".type        postmortem_kernel_exception_handler, @function\n\t"
+    "\n"
+"postmortem_kernel_exception_handler:\n\t" // Copy destination _KernelExceptionVector
+    "wsr.excsave1 a0\n\t"
+
+"postmortem_kernel_exception_handler_p3:\n\t"
+    "addi         a0,     a1,     -32\n\t"
+    "s32i         a1,     a0,     24\n\t"
+    "mov.n        a1,     a0\n\t"
+    "xsr.excsave2 a0\n\t"                 // stash pointer to stack data in excsave2
+    "s32i         a0,     a1,     12\n\t" // and save previous
+    "rsr.excsave1 a0\n\t"
+    "s32i         a0,     a1,     28\n\t"
+    "rsr.epc1     a0\n\t"
+    "s32i         a0,     a1,     20\n\t"
+    // save current exccause
+    "rsr.exccause a0\n\t"
+    "j            . - 54\n\t"   // continue at postmortem_debug_exception_vector_part2
+
+    ".size postmortem_kernel_exception_handler, .-postmortem_kernel_exception_handler\n\t"
+);
+
+void postmortem_double_exception_handler(void);
+asm(  // 16 byte MAX, using 14
+    ".section     .text.postmortem_double_exception_handler,\"ax\",@progbits\n\t"
+    ".align       4\n\t"
+    ".type        postmortem_double_exception_handler, @function\n\t"
+    "\n"
+"postmortem_double_exception_handler:\n\t" // Copy destination _DoubleExceptionVector
+    // Common code at _KernelExceptionVector, which is -64 bytes from
+    // _DoubleExceptionVector.
+    "j             .  - 64\n\t" // continue at _KernelExceptionVector
+
+    // finish here from _DebugExceptionVector part 2
+    // _DoubleExceptionVector part 2
+"postmortem_double_exception_handler_part2:\n\t"
+    // User mode, Clear EXCM, and block all interrupts
+    "movi.n       a0,    0x02F\n\t"
+    "wsr.ps       a0\n\t"
+    "rsync\n\t"
+    // Let _UserExceptionVector finish processing the exception as an
+    // exception(0). Adjust results at __wrap_system_restart_local so that `ESP
+    // Exception Decoder` reports the state before this `ill` instruction.
+    "ill\n\t"
+    ".size postmortem_double_exception_handler, .-postmortem_double_exception_handler\n\t"
+);
+
+/*
+  Of the Exception Cause table's 64 entries, most are set to
+  `_xtos_unhandled_exception`. And I assume most of these never occur; however,
+  at lesat one does exccause 20. This one occurs on a call to a NULL pointer. This
+  can happen when a function prototype is given the weak attribute but never
+  defined, or forgetting to test a call back function pointer for NULL before
+  calling, etc.
+
+  A problem with the `_xtos_unhandled_exception` function is that it handles the
+  event with a Debug Exception BP. If you don't have GDB running you see a HWDT
+  reset. If interrupts are at INTLEVEL 2 or higher (a BP instruction becomes a
+  NOP), you still see a HWDT reset regardless of running GDB.
+*/
+typedef void (* _xtos_handler)(void);
+static  _xtos_handler * const _xtos_exc_handler_table = (_xtos_handler *)0x3FFFC000u;
+#define ROM_xtos_unhandled_exception (reinterpret_cast<_xtos_handler>(0x4000dc44))
```

Comment:
What is this magic address? Is it a ROM function? Does it make sense to move to the file where the ROM functions are mapped to addresses?
If not, does it make sense to change this define into a constexpr with a type?

Reply:
While `_xtos_unhandled_exception` is in the linker `.ld` file, it may have been overridden by the sketch or future SDK(not likely). We require the original Boot ROM function address to limit our override to those old values in the table.

With the new toolchain change `constexpr` doesn't work for this. It does not like an integer being cast to a function:
```cpp
constexpr _xtos_handler ROM_xtos_unhandled_exception = (reinterpret_cast<_xtos_handler>(0x4000dc44));

error: 'reinterpret_cast' from integer to pointer
```
It works if I replace `constexpr` with `const` but I was concerned it would take up DRAM. I just tried it and in this case, it doesn't appear to change DRAM usage.

I'll change it to `const` for now.